Performance guidance grammar/length cleanup (#1737)
MicrosoftDocs/mixed-reality: mixed-reality-docs/understanding-performance-for-mixed-reality.md

# Understanding performance for mixed reality

This article is an introduction to understanding the significance of performance for your Mixed Reality app.  User experience can be greatly degraded if your application does not run at optimal frame rate. Holograms will appear unstable and head tracking of the environment will be inaccurate, leading to an poor experience for the user. Performance must be considered a first class feature for mixed reality development and not a polish task.

The performant framerate values for each target platform are listed below.

| Platform | Target Frame Rate |
|----------|-------------------|
| [HoloLens](hololens-hardware-details.md) | 60 FPS |
| [Windows Mixed Reality Ultra PCs](immersive-headset-hardware-details.md) | 90 FPS |
| [Windows Mixed Reality PCs](immersive-headset-hardware-details.md) | 60 FPS |

> The framework below outlines best practices for hitting target frame rates. If you're developing in Unity, consider reading the [performance recommendations for Unity article](performance-recommendations-for-unity.md) as well for tips on measuring and improving your framerate in the Unity environment.

Reviewer: ```suggestion
The framework below outlines best practices for hitting target frame rates. If developing in Unity, consider reading the [performance recommendations for Unity article](performance-recommendations-for-unity.md) as well for tips on measuring and improving framerate in the Unity environment.
```

1. **App Thread - CPU** -
    This thread is responsible for your app logic. This includes processing input, animations, physics, and other app logic.
2. **Render Thread - CPU to GPU** - 
    This thread is responsible for submitting your draw calls to the GPU. When your app wants to render an object such as a cube or model, this thread sends a request to the GPU to perform these operations.
3. **GPU** - 
    This processor most commonly handles the graphics pipeline of your application to transform 3D data (models, textures, etc.) into pixels. It ultimately produces a 2D image to submit to your device's screen.

![Lifetime of a Frame](images/lifetime-of-a-frame.png)

Generally, HoloLens applications will be GPU bound, but not always. Use the tools and techniques below to understand where your particular app is bottlenecked.

## How to analyze your application

> There are many tools that allow you to understand the performance profile of your mixed reality application. These will enable you find where and why you have bottlenecks so you can address them.

Reviewer: ```suggestion
There are many tools that allow you to understand the performance profile of your mixed reality application. These will enable you to find where and why you have bottlenecks so you can address them.
```

Below are some common tools to gain deep profiling information for your application:
- [Intel Graphics Performance Analyzers](https://software.intel.com/gpa)
- [Visual Studio Graphics Debuggers](https://docs.microsoft.com/visualstudio/debugger/graphics/visual-studio-graphics-diagnostics?view=vs-2017)
- [Unity Profiler](https://docs.unity3d.com/Manual/Profiler.html)
- [Unity Frame Debugger](https://docs.unity3d.com/Manual/FrameDebugger.html)

### How to profile in any environment

> One way to determine if you are GPU bound or CPU bound in your application is to decrease the resolution of the render target output. By reducing the number of pixels to calculate, this will reduce your GPU load. The device will render to a smaller surface, then up-sample to display your final image.

Reviewer: ```suggestion
One way to determine if you are GPU bound or CPU bound in your application is to decrease the resolution of the render target output. By reducing the number of pixels to calculate, this will reduce your GPU load. The device will render to a smaller texture, then up-sample to display your final image.
```

    - To identify fill rate limitations, decrease the display resolution and check if framerate improved. 
    - In Unity, this can be done via the  *[XRSettings.renderViewportScale](https://docs.unity3d.com/ScriptReference/XR.XRSettings-renderViewportScale.html)* property

Memory bandwidth generally involves optimizations to either:
1) Decrease texture resolutions
2) Utilize fewer textures (normals, specular, etc.)

Fill rate is focused on reducing the number of operations that need to be computed for a final rendered pixel. This includes reducing:
1) Number of objects to render/process
2) Number of operations per shader
3) Number of GPU stages to final result (geometry shaders, post-processing effects, etc.)
4) Number of pixels to render (display resolution)

#### Reduce polygon count
Higher polygon counts result in more operations for the GPU; reducing the number of polygons in your scene will reduce the render time. There are other factors involved in shading the geometry that can be expensive, but polygon count is the simplest metric to determine how expensive a scene will be to render.

#### Limit overdraw

> High overdraw occurs when multiple objects are rendered but not output to the screen as they are hidden by an occluding object. Imagine looking at a wall that has objects behind it. All of the geometry would be processed for rendering, but only the opaque wall needs to be rendered. This results in unnecessary operations.

Reviewer: ```suggestion
High overdraw occurs when multiple objects are rendered but not shown on screen as they are hidden by an occluding object. Imagine looking at a wall that has objects behind it. All of the geometry would be processed for rendering, but only the opaque wall needs to be rendered. This results in unnecessary operations.
```

#### Shaders

Shaders are small programs that run on the GPU and perform two important steps in rendering:
1) Determining which vertices should be drawn and where they are in screen space (the Vertex shader)
    - The Vertex shader is generally executed per vertex for every mesh.
2) Determining the color of each pixel (the Pixel shader)
    - The Pixel shader is executed per pixel rendered by the geometry to the texture being rendered to.

Typically shaders perform many transformations and lighting calculations. Although complex lighting models, shadows, and other operations can generate fantastic results, they also come with a price. Reducing the number of operations computed in shaders can greatly reduce the work needed for the GPU per frame.

##### Shader coding recommendations

- Use bilinear filtering whenever possible
- Rearrange expressions to use MAD intrinsics in order to do a multiply and an add at the same time
- Precalculate as much as possible on the CPU and pass as constants to the material
- **Favor moving operations from the pixel shader to the vertex shader**

>     - Generally the number of vertices is much smaller than the number of pixels (720p == 921,600 pixels, 1080p == 2,073,600 pixels, etc.)

Reviewer: ```suggestion
    - Generally the number of vertices is much smaller than the number of pixels (720p is 921,600 pixels, 1080p is 2,073,600 pixels, etc.)
```
